### PR TITLE
make `spack fetch` work with environments

### PR DESCRIPTION
**Pull request by healther**
I was trying to setup shop on a new supercomputer and since I'm using environments now on my own machine I thought I could just use that. Turns out you need to fetch things from the login node, rather than do that on the compute node...

Had to teach `spack fetch` to gather its todo list from an environment. Let me know if you want to keep this 

* previously: `spack fetch` required the explicit statement of
              the specs to be fetched, even when in an environment
* now: if there is no spec(s) provided to `spack fetch` we check
       if an environment is active and if yes we fetch all
       uninstalled specs.

### CODE REVIEW

**Comment by healther** [comment]
I don't think I broke that?

**Comment by adamjstewart** [comment]
I like this feature. I restarted the failing tests, I think it was just apt problems.

**Comment by healther** [comment]
> ==> Error: InstallError: mpich +fortran requires Fortran compilers. Configure Fortran compiler or disable Fortran support with ~fortran

I'm compartatively certain I didn't break that with this PR ;)

What I'm not sure about is the semantics of being in an environment and providing an explicit spec. I think the one implemented here is the only sensible one, but it might be that we want to actively disallow a manual spec provision within an environment (not sure how it is handled in other cases)

**Comment by adamjstewart** [comment]
> What I'm not sure about is the semantics of being in an environment and providing an explicit spec. I think the one implemented here is the only sensible one, but it might be that we want to actively disallow a manual spec provision within an environment (not sure how it is handled in other cases)

For other commands inside environments, I'm actually disallowed from running commands on packages that are not in my spec. I don't actually like this behavior, so I prefer yours, but if other contributors think that consistency is more important I'll let them decide.

**Comment by healther** [comment]
Is it useful/necessary to test the `cmd/fetch.py` file? If yes, what would you like to have tested?

**Comment by adamjstewart** [comment]
It's probably difficult to test `cmd/fetch.py` since it requires internet access, but maybe @alalazo has some ideas.

**Comment by healther** [comment]
Then someone from the higher ups: Can this stay as is? (I'd slightly prefer not to have to fiddle with unittest-mock-internet-things, on the other hand, I'd rather have my fetch method being upstream...)

**Review by adamjstewart — DISMISSED** [review]

**Comment by adamjstewart** [comment]
Can you resolve the merge conflict?

**Comment by healther** [comment]
Done, while running the tests (to see if I broke anything) I noticed that they take quite a while... 

```
======================================================================== slowest 20 test durations ========================================================================
359.94s call     lib/spack/spack/test/cmd/license.py::test_list_files
41.18s call     lib/spack/spack/test/cmd/url.py::test_url_list
37.59s call     lib/spack/spack/test/cmd/clean.py::test_function_calls[-a-counters5]
17.31s call     lib/spack/spack/test/directory_layout.py::test_yaml_directory_layout_parameters
15.94s call     lib/spack/spack/test/bindist.py::test_default_rpaths_create_install_default_layout
13.59s call     lib/spack/spack/test/cmd/env.py::test_stack_combinatorial_view[mock_archive0]
13.29s call     lib/spack/spack/test/package_sanity.py::test_get_all_packages
12.66s call     lib/spack/spack/test/cmd/list.py::test_list_format_html
11.95s call     lib/spack/spack/test/bindist.py::test_default_rpaths_install_nondefault_layout
11.28s call     lib/spack/spack/test/cmd/env.py::test_stack_view_multiple_views[mock_archive0]
10.87s call     lib/spack/spack/test/cmd/env.py::test_stack_view_select_and_exclude[mock_archive0]
10.78s call     lib/spack/spack/test/cmd/env.py::test_stack_view_activate_from_default[mock_archive0]
10.43s call     lib/spack/spack/test/bindist.py::test_relative_rpaths_create_default_layout
10.31s call     lib/spack/spack/test/cmd/env.py::test_view_link_roots[mock_archive0]
10.12s call     lib/spack/spack/test/cmd/env.py::test_stack_view_exclude[mock_archive0]
9.79s call     lib/spack/spack/test/cmd/env.py::test_stack_view_select[mock_archive0]
9.68s call     lib/spack/spack/test/cmd/env.py::test_stack_view_no_activate_without_default[mock_archive0]
9.07s call     lib/spack/spack/test/cmd/env.py::test_view_link_all[mock_archive0]
9.02s call     lib/spack/spack/test/cmd/url.py::test_url_stats
9.02s call     lib/spack/spack/test/cmd/url.py::test_url_summary
```

To be fair this is on a somewhat old MacBook Air, but still... 6 minutes for checking the licenses, has anyone taken a look at speeding things up?

And I'm pretty sure the fail of
```
FAIL lib/spack/spack/test/bindist.py::test_push_and_fetch_keys
```
should have been an XFAIL or SKIP right?

**Comment by alalazo** [comment]
@healther I think we accumulated over time tests that are more similar to e2e or integration tests rather than unit tests, in particular for the binary packaging part of Spack. Should we discuss this topic in a separate issue?

**Comment by adamjstewart** [comment]
I vote we merge as is, but I'm waiting for one more approval from a core developer.

**Comment by healther** [comment]
@alalazo I didn't want to change this in this PR, just remarking on the ... resource requirements of the `spack test` command :) I'd merge this as is and at some point in the future would like to setup some kind of "reduce testing CPU wastage" PR. 

Edit. There you go: #19369

**Comment by becker33** [comment]
@healther could you add a test using the `mock_fetch` fixture in pytest to avoid actual dependencies on the internet?

**Comment by healther** [comment]
I'll take a look if I can squeeze it in, but I don't have much experience with mocking, so it will probably take me a while... If you've got the experience (and the time) you should be able to push to this PR right?

**Review by tgamblin — CHANGES_REQUESTED** [review]
@healther: I'd like to get this in but it needs better test coverage.  Can you add tests to cover the patch?

**Comment by healther** [comment]
> @healther: I'd like to get this in but it needs better test coverage. Can you add tests to cover the patch?

@tgamblin I'll have a look this weekend/early next week. Had my defence last week, so effectively canceled everything else xD I'll ask in slack if I need pointers regarding the mocking stuff

**Comment by healther** [comment]
@tgamblin @becker33 I'm not sure if this is what you had in mind for tests, but that's what I could come up with. It seems to me that the whole fetching thing is not really well covered... Let me know if you have better/more ideas.

File: lib/spack/spack/test/cmd/fetch.py
Comment (adamjstewart) [inline lib/spack/spack/test/cmd/fetch.py]: Shouldn't `spack fetch` in an environment download all of the packages in that environment? Why does it raise an error?

File: lib/spack/spack/cmd/fetch.py
Comment (adamjstewart) [inline lib/spack/spack/cmd/fetch.py]: I think we should fetch all packages in the environment, not just uninstalled specs. If they are already present in the cache, they will be skipped automatically.

File: lib/spack/spack/cmd/fetch.py
Comment (adamjstewart) [inline lib/spack/spack/cmd/fetch.py:43]: Other Spack commands will fail when you try to fetch a package that is not present in an environment. Should we do the same thing here?

File: lib/spack/spack/cmd/fetch.py
Comment (adamjstewart) [inline lib/spack/spack/cmd/fetch.py]:
```suggestion
            tty.die("fetch requires at least one spec argument")
```

File: lib/spack/spack/test/cmd/fetch.py
Comment (adamjstewart) [inline lib/spack/spack/test/cmd/fetch.py]: Is it possible to add a mark that specifies that the following tests require internet access? Do we have such a mark yet?

File: lib/spack/spack/cmd/fetch.py
Comment (healther) [reply lib/spack/spack/cmd/fetch.py]: Are you sure? Fetching provides the sources, which may no longer be around once things are installed. And we do not need them as we won't be building those anymore. I don't really see an upside to just fetching all

File: lib/spack/spack/cmd/fetch.py
Comment (healther) [reply lib/spack/spack/cmd/fetch.py:43]: What would be the upside? I guess it would more clearly distinguish between in-environment and out-of-environment behaviour, but if I say `spack fetch <package>` I simply expect spack to fetch the thing and not annoy me with "technically this isn't in the environment"

File: lib/spack/spack/test/cmd/fetch.py
Comment (healther) [reply lib/spack/spack/test/cmd/fetch.py]: I do not really know anything about this... I do not thing any of this here requires internet per se, but I haven't explicitly checked and I barely know my way around basic pytest let alone the collection of fixtures used here

File: lib/spack/spack/cmd/fetch.py
Comment (healther) [reply lib/spack/spack/cmd/fetch.py]: Done

**Comment by healther** [comment]
Is the line width of 80 something that we are married to? `black` defaults to 120 (it's the one configuration that can be overridden, so it's fine, but still annoying :))

**Comment by adamjstewart** [comment]
Black defaults to 88, not 120. I think we're married to 79 for now, but once we switch to black I imagine we'll switch to 88.

File: lib/spack/spack/cmd/fetch.py
Comment (tgamblin) [reply lib/spack/spack/cmd/fetch.py:43]: I agree with @healther on this one -- `fetch` doesn't affect environment state and shouldn't be affected by it.

File: lib/spack/spack/cmd/fetch.py
Comment (tgamblin) [reply lib/spack/spack/cmd/fetch.py]: I think the usual reason someone will try to fetch an entire environment (at least at our site) is to get all the sources so that they can build it somewhere else.  They may have installed some things *right here* but they probably want to build those someplace else too.

Can you make this fetch everything by default?  IMO, the end state of the operation should by default be idempotent -- I run `spack fetch` in an environment, and I have everything I need to build it.

Fetching uninstalled can be an option as I can see how that's useful but I do think it's important that you can say "I ran spack fetch and therefore I could rebuild all of this".

**Review by tgamblin — CHANGES_REQUESTED** [review]
Thanks for the test!  One last change request (I agree w/@adamjstewart) and we're good to go.

**Comment by healther** [comment]
Done, although I'd argue that the default should be the other way around, but I didn't care enough to change the default based on whether we are in an environment or not. Hope this is what you had in mind.

**Comment by healther** [comment]
I was looking for a way to document the environment specific behaviour, either as 
- an additional description line, but this shows up in `spack help --all`
- an extension of the specs description, but this would affect all others and in particular something like `spack remove` does not just take "all specs from the environment" if the argument is empty.

Not sure how we could fit that in elegantly, so I kinda gave up, if you have brilliant ideas just shout them out, otherwise I'd say merge as is, and when I'm bored again I'll take a closer look at the help formatting system we use.

**Comment by adamjstewart** [comment]
Try `parser.epilog`. See `spack build-env` for an example.

**Comment by healther** [comment]
> Try `parser.epilog`. See `spack build-env` for an example.

Thanks, that worked. Unfortunately black objected to the formatting of the other parser arguments (when I had it reorganise the epilog string). I tend to just take whatever black produces so I just committed it. If you want me to revert the drivebys, let me know (or if you can do it directly in the merge feel free to)

**Comment by healther** [comment]
and rebased, someone double check that i didn't break anything, but I think the conflict resolution was straight forward

File: lib/spack/spack/test/cmd/fetch.py
Comment (tgamblin) [reply lib/spack/spack/test/cmd/fetch.py]: Yeah we should make these tests not use the internet.  I can try to rework this.  I think we can leave @healther to his work at this point 😄

File: lib/spack/spack/test/cmd/fetch.py
Comment (healther) [reply lib/spack/spack/test/cmd/fetch.py]: Because at this point we don't know which packages are in the environment at all yet (didn't do concretize yet)

**Comment by tldahlgren** [comment]
Closing and re-opening this PR to force required tests to complete/report results.

**Comment by tldahlgren** [comment]
@tgamblin Are you satisfied with the change?

**Comment by tgamblin** [comment]
The tests here still need to be completely local, otherwise they won’t pass on an air gap network. We can move this to 0.16.2.

File: lib/spack/spack/cmd/fetch.py
Comment (becker33) [inline lib/spack/spack/cmd/fetch.py:57]: I would think this PR should concretize the environment, the same way that the `spack fetch` command currently concretizes specs.

**Review by tgamblin — DISMISSED** [review]

**Review by tgamblin — APPROVED** [review]
These tests work without network access, so LGTM.  Thanks @healther!  I think I earlier missed that this had been pushed and that it worked.

**Comment by haampie** [comment]
The feature is great, the tests are ... not testing anything?

**Comment by haampie** [comment]
Once more I'm hitting this PR due to an issue :|.

@tgamblin: can we at the very least prefix `uninstalled_specs` with a `_` to signal it's private API?

The naming is bad, and it's confusing. It returns the _root_ specs that are not installed, _or_ are installed but are develop specs, _or_ have develop specs among their dependencies...

This was definitely merged too soon.

See https://github.com/spack/spack/pull/31643

**Pull request merged**
